pkg/agentprofile/agent_profile.go

```diff
@@ -29,16 +32,52 @@ const (
 // - If there are no conflicting profiles, it returns all the profiles plus the default one.
 // - If there are conflicting profiles, it returns a subset that does not
 // conflict plus the default one. When there are conflicting profiles, the
-// oldest one is the one that takes precedence.
-func ProfilesToApply(profiles []datadoghqv1alpha1.DatadogAgentProfile) []datadoghqv1alpha1.DatadogAgentProfile {
+// oldest one is the one that takes precedence. When two profiles share an
+// identical creation timestamp, the profile whose name is alphabetically first
+// is considered to have priority.
+func ProfilesToApply(profiles []datadoghqv1alpha1.DatadogAgentProfile, nodes []v1.Node) ([]datadoghqv1alpha1.DatadogAgentProfile, error) {
 	var res []datadoghqv1alpha1.DatadogAgentProfile
 
-	// TODO: detect conflicts here and only add the ones that are not
-	// conflicting (Give precedence to the oldest profile).
-	// This function will need the list of hosts to check for conflicts.
-	res = append(res, profiles...)
+	nodesWithProfilesApplied := make(map[string]bool, len(nodes))
+	for _, node := range nodes {
+		nodesWithProfilesApplied[node.Name] = false
+	}
+
+	sortedProfiles := sortProfiles(profiles)
+
+	for _, profile := range sortedProfiles {
+		conflicts := false
+		nodesThatMatchProfile := map[string]bool{}
 
-	return append(res, defaultProfile(res))
+		for _, node := range nodes {
+			matchesNode, err := profileMatchesNode(&profile, &node)
+			if err != nil {
+				return nil, err
+			}
+
+			if matchesNode {
+				if nodesWithProfilesApplied[node.Name] {
+					// Conflict. This profile should not be applied.
```

Comment:
It would be nice if we could surface profile conflicts in the operator logs so that users know that a conflicting profile was found and not applied. Would this be the right place to log a message?

Reply:
I think that for users it would be better show this kind of information in the status of the profile resource.
We don't have anything yet for the status, so I prefer to leave it for later in a separate PR.